Reserve before `write_fmt` for owned buffers

thaliaarchi opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/137762)*
<!-- homu-ignore:end -->

`fmt::Arguments::estimated_capacity()` is currently only used by `fmt::format()` to reserve the initial string capacity. Also use it in `fmt::Write::write_fmt` for `String` and `OsString`; and `io::Write::write_fmt` for `Vec<u8>`, `VecDeque<u8>`, `Cursor<&mut Vec<u8>>`, and `Cursor<Vec<u8>>`.

This may be worth checking perf.

thaliaarchi commented:
r? @workingjubilee

workingjubilee commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 0e82180c453e9f97da818fd215d817de2e030f25 with merge 9cdfd5df41f0c63cf60717083c177a4af5071687...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0e82180c453e9f97da818fd215d817de2e030f25","merge_sha":"9cdfd5df41f0c63cf60717083c177a4af5071687"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13579135379/job/37965446950)
Build commit: 9cdfd5df41f0c63cf60717083c177a4af5071687 (`9cdfd5df41f0c63cf60717083c177a4af5071687`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13579135379/job/37965446950"},"merge_sha":"9cdfd5df41f0c63cf60717083c177a4af5071687"} -->

rust-timer commented:
Queued 9cdfd5df41f0c63cf60717083c177a4af5071687 with parent f45d4acf1bb635aa010f19f8a749eed8293203b3, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f45d4acf1bb635aa010f19f8a749eed8293203b3&end=9cdfd5df41f0c63cf60717083c177a4af5071687).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (9cdfd5df41f0c63cf60717083c177a4af5071687): [comparison URL](https://perf.rust-lang.org/compare.html?start=f45d4acf1bb635aa010f19f8a749eed8293203b3&end=9cdfd5df41f0c63cf60717083c177a4af5071687&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=f45d4acf1bb635aa010f19f8a749eed8293203b3&end=9cdfd5df41f0c63cf60717083c177a4af5071687&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f45d4acf1bb635aa010f19f8a749eed8293203b3&end=9cdfd5df41f0c63cf60717083c177a4af5071687&stat=max-rss)
<details>
<summary>Results (primary 10.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 10.1% | [10.1%, 10.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 10.1% | [10.1%, 10.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f45d4acf1bb635aa010f19f8a749eed8293203b3&end=9cdfd5df41f0c63cf60717083c177a4af5071687&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=f45d4acf1bb635aa010f19f8a749eed8293203b3&end=9cdfd5df41f0c63cf60717083c177a4af5071687&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.2%]   | 10    |
| Regressions ❌ <br /> (secondary)  | 0.1%  | [0.1%, 0.2%]   | 37    |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.0%] | 7     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.0%  | [-0.2%, 0.2%]  | 17    |
</details>

**Bootstrap**: 770.19s -> 771.304s (0.14%)
**Artifact size**: 362.00 MiB -> 361.96 MiB (-0.01%)

thaliaarchi commented:
No regressions, so are we good?

thaliaarchi committed:
Reserve before write_fmt for owned buffers

Reserve before formatting with `fmt::Arguments::estimated_capacity()` in
`fmt::Write::write_fmt` and `io::Write::write_fmt` implementations for
owned buffer types.

Adding `#[inline]` to `write_fmt` shows minor perf regressions, so leave
it off like the default impl.

thaliaarchi commented:
I think the `fmt::Write` side is more likely to have an impact. Let’s see.

@bors try @rust-timer queue

bors commented:
@thaliaarchi: :key: Insufficient privileges: not in try users

rust-timer commented:
Insufficient permissions to issue commands to rust-timer.
<!-- rust-timer: temporary -->

Noratrieb commented:
rustdoc uses a lot of formatting, so this might have a positive impact there
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit d8b6fe87686995895e7f8557fd9cae19fb9cfae0 with merge ba1e46459f988e4c6b39e6b34e656327b9e9fcd9...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d8b6fe87686995895e7f8557fd9cae19fb9cfae0","merge_sha":"ba1e46459f988e4c6b39e6b34e656327b9e9fcd9"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13594851753/job/38014615367)
Build commit: ba1e46459f988e4c6b39e6b34e656327b9e9fcd9 (`ba1e46459f988e4c6b39e6b34e656327b9e9fcd9`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13594851753/job/38014615367"},"merge_sha":"ba1e46459f988e4c6b39e6b34e656327b9e9fcd9"} -->

rust-timer commented:
Queued ba1e46459f988e4c6b39e6b34e656327b9e9fcd9 with parent 60493b8973ac5ba632952eaa2f212b56bb97ccfe, future [comparison URL](https://perf.rust-lang.org/compare.html?start=60493b8973ac5ba632952eaa2f212b56bb97ccfe&end=ba1e46459f988e4c6b39e6b34e656327b9e9fcd9).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (ba1e46459f988e4c6b39e6b34e656327b9e9fcd9): [comparison URL](https://perf.rust-lang.org/compare.html?start=60493b8973ac5ba632952eaa2f212b56bb97ccfe&end=ba1e46459f988e4c6b39e6b34e656327b9e9fcd9&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=60493b8973ac5ba632952eaa2f212b56bb97ccfe&end=ba1e46459f988e4c6b39e6b34e656327b9e9fcd9&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.2%, 1.1%]   | 13    |
| Regressions ❌ <br /> (secondary)  | 0.7%  | [0.7%, 0.7%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.7%, -0.3%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.5%  | [-0.7%, 1.1%]  | 15    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=60493b8973ac5ba632952eaa2f212b56bb97ccfe&end=ba1e46459f988e4c6b39e6b34e656327b9e9fcd9&stat=max-rss)
<details>
<summary>Results (primary 2.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.6%  | [1.1%, 12.3%]  | 6     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-3.1%, -1.9%] | 3     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 2.8%  | [-3.1%, 12.3%] | 9     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=60493b8973ac5ba632952eaa2f212b56bb97ccfe&end=ba1e46459f988e4c6b39e6b34e656327b9e9fcd9&stat=cycles:u)
<details>
<summary>Results (primary 0.9%, secondary -2.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9%  | [0.7%, 1.0%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-2.4%, -2.4%] | 1     |
| All ❌✅ (primary)                 | 0.9%  | [0.7%, 1.0%]   | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=60493b8973ac5ba632952eaa2f212b56bb97ccfe&end=ba1e46459f988e4c6b39e6b34e656327b9e9fcd9&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.1%, secondary 0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.0%, 0.3%]   | 62    |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.2%, 0.7%]   | 40    |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.5%, -0.0%] | 10    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-0.5%, 0.3%]  | 72    |
</details>

**Bootstrap**: 770.987s -> 775.043s (0.53%)
**Artifact size**: 361.90 MiB -> 361.90 MiB (0.00%)

workingjubilee commented:
that is very interesting. some size and icount regressions suggests this pushes more code into hot paths. I can accept this with the last commit reverted (well, rebased away, whatever).

thaliaarchi commented:
> some size and icount regressions suggests this pushes more code into hot paths. I can accept this with the last commit reverted

I want to experiment more first, but if it still regresses, I'll drop the `fmt::Write` commit. I think there are two things to try:
1. Remove the new `#[inline]` annotations. The default impl doesn't have them.
2. Only do `self.reserve(args.estimated_capacity())` when the args are statically known. `estimated_capacity` is constant propagated by the compiler when known, but maybe it doesn't perform well otherwise?

workingjubilee commented:
Happy to continue putting this into the perf queue for a bit!

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 167798ed1615363e9f49a01d24810a8ef1c1375d with merge 3bdd33d1925705642d613bc6f77284a2be53eb5d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"167798ed1615363e9f49a01d24810a8ef1c1375d","merge_sha":"3bdd33d1925705642d613bc6f77284a2be53eb5d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13599529507/job/38025835983)
Build commit: 3bdd33d1925705642d613bc6f77284a2be53eb5d (`3bdd33d1925705642d613bc6f77284a2be53eb5d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13599529507/job/38025835983"},"merge_sha":"3bdd33d1925705642d613bc6f77284a2be53eb5d"} -->

rust-timer commented:
Queued 3bdd33d1925705642d613bc6f77284a2be53eb5d with parent aa3c2d73eff57d6b019c26a1e9aa8afd8bc186ad, future [comparison URL](https://perf.rust-lang.org/compare.html?start=aa3c2d73eff57d6b019c26a1e9aa8afd8bc186ad&end=3bdd33d1925705642d613bc6f77284a2be53eb5d).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (3bdd33d1925705642d613bc6f77284a2be53eb5d): [comparison URL](https://perf.rust-lang.org/compare.html?start=aa3c2d73eff57d6b019c26a1e9aa8afd8bc186ad&end=3bdd33d1925705642d613bc6f77284a2be53eb5d&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=aa3c2d73eff57d6b019c26a1e9aa8afd8bc186ad&end=3bdd33d1925705642d613bc6f77284a2be53eb5d&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.3%, 0.7%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 0.6%  | [0.3%, 0.8%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.7%, -0.3%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.1% | [-0.7%, 0.7%]  | 6     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=aa3c2d73eff57d6b019c26a1e9aa8afd8bc186ad&end=3bdd33d1925705642d613bc6f77284a2be53eb5d&stat=max-rss)
<details>
<summary>Results (primary 1.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 10.1% | [10.1%, 10.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.3% | [-3.1%, -1.6%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.8%  | [-3.1%, 10.1%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=aa3c2d73eff57d6b019c26a1e9aa8afd8bc186ad&end=3bdd33d1925705642d613bc6f77284a2be53eb5d&stat=cycles:u)
<details>
<summary>Results (secondary 3.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.0% | [1.4%, 4.0%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=aa3c2d73eff57d6b019c26a1e9aa8afd8bc186ad&end=3bdd33d1925705642d613bc6f77284a2be53eb5d&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary 0.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.0%, 0.5%]   | 9     |
| Regressions ❌ <br /> (secondary)  | 0.5%  | [0.3%, 0.5%]   | 35    |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.8%, -0.0%] | 22    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-0.8%, 0.5%]  | 31    |
</details>

**Bootstrap**: 770.727s -> 771.43s (0.09%)
**Artifact size**: 361.98 MiB -> 361.95 MiB (-0.01%)

workingjubilee commented:
Kinda weird that it's still mixed overall, but yeah, less inlining is much better.

workingjubilee commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit f5f8559f4de0a4b30b10a005e746a4efc5d8b473 with merge e1eb8ce1e002740b2b820f343fb6642f0afe2a10...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f5f8559f4de0a4b30b10a005e746a4efc5d8b473","merge_sha":"e1eb8ce1e002740b2b820f343fb6642f0afe2a10"} -->

thaliaarchi commented:
@workingjubilee I neglected to fix/disable a test which was testing the old behavior of `estimated_capacity`, so the test failed. We need a new perf run.

Dylan-DPC commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 9c2a9ae35f4be1247731ee47a96427548e8c6193 with merge 92eb78b2fce9c9ed65d541f2fa41e63f777dded5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9c2a9ae35f4be1247731ee47a96427548e8c6193","merge_sha":"92eb78b2fce9c9ed65d541f2fa41e63f777dded5"} -->

Noratrieb commented:
why would you need a new perf run? the runtime or compile time performance doesn't change when disabling a test.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13608000375/job/38043623367)
Build commit: 92eb78b2fce9c9ed65d541f2fa41e63f777dded5 (`92eb78b2fce9c9ed65d541f2fa41e63f777dded5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13608000375/job/38043623367"},"merge_sha":"92eb78b2fce9c9ed65d541f2fa41e63f777dded5"} -->

rust-timer commented:
Queued 92eb78b2fce9c9ed65d541f2fa41e63f777dded5 with parent 8c392966a013fd8a09e6b78b3c8d6e442bc278e1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=8c392966a013fd8a09e6b78b3c8d6e442bc278e1&end=92eb78b2fce9c9ed65d541f2fa41e63f777dded5).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (92eb78b2fce9c9ed65d541f2fa41e63f777dded5): [comparison URL](https://perf.rust-lang.org/compare.html?start=8c392966a013fd8a09e6b78b3c8d6e442bc278e1&end=92eb78b2fce9c9ed65d541f2fa41e63f777dded5&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=8c392966a013fd8a09e6b78b3c8d6e442bc278e1&end=92eb78b2fce9c9ed65d541f2fa41e63f777dded5&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.3%, 0.6%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 0.1%  | [0.1%, 0.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.4%, -0.2%] | 3     |
| Improvements ✅ <br /> (secondary) | -0.5% | [-0.7%, -0.3%] | 13    |
| All ❌✅ (primary)                 | 0.0%  | [-0.4%, 0.6%]  | 5     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=8c392966a013fd8a09e6b78b3c8d6e442bc278e1&end=92eb78b2fce9c9ed65d541f2fa41e63f777dded5&stat=max-rss)
<details>
<summary>Results (primary 2.4%, secondary 5.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 7.4%  | [0.8%, 14.0%]  | 2     |
| Regressions ❌ <br /> (secondary)  | 5.3%  | [5.1%, 5.5%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-4.7%, -0.6%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 2.4%  | [-4.7%, 14.0%] | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=8c392966a013fd8a09e6b78b3c8d6e442bc278e1&end=92eb78b2fce9c9ed65d541f2fa41e63f777dded5&stat=cycles:u)
<details>
<summary>Results (primary -0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.4%  | [1.4%, 1.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.0% | [-2.0%, -2.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.3% | [-2.0%, 1.4%]  | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=8c392966a013fd8a09e6b78b3c8d6e442bc278e1&end=92eb78b2fce9c9ed65d541f2fa41e63f777dded5&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.0%, 0.5%]   | 13    |
| Regressions ❌ <br /> (secondary)  | 0.5%  | [0.3%, 0.5%]   | 35    |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.3%, -0.0%] | 17    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.0%  | [-0.3%, 0.5%]  | 30    |
</details>

**Bootstrap**: 773.316s -> 769.588s (-0.48%)
**Artifact size**: 361.90 MiB -> 361.90 MiB (0.00%)

workingjubilee commented:
I'm not entirely sure what you changed if anything in your last rebase, @thaliaarchi?

thaliaarchi commented:
Since I rebased onto master, the changes got muddled; sorry. You can use `git range diff before...after` to nicely compare rebases. The change was to use `as_statically_known_str` in `io::Write::write_fmt` like `fmt::Write::write_fmt` already does:

```diff
-    #[inline]
-    fn write_fmt(&mut self, fmt: fmt::Arguments<'_>) -> io::Result<()> {
-        self.reserve(fmt.estimated_capacity());
-        io::default_write_fmt(self, fmt)
+    fn write_fmt(&mut self, args: fmt::Arguments<'_>) -> io::Result<()> {
+        if let Some(s) = args.as_statically_known_str() {
+            self.write_all(s.as_bytes())
+        } else {
+            self.reserve(args.estimated_capacity());
+            io::default_write_fmt(self, args)
+        }
     }
```

I'm marking this PR as a draft for now, pending my investigation into a better heuristic for `fmt::Arguments::estimated_capacity`. My WIP hook for gathering stats is in my [format-args-stats](https://github.com/thaliaarchi/rust/commits/format-args-stats) branch.

thaliaarchi commented:
Now that https://github.com/rust-lang/rust/pull/138650 has merged, so the non-`reserve` part of this PR is in master, I want to see more clearly where we're at. Can we do another perf run?

Noratrieb commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit a1c03f31925da058f5bbbb3b2d8e00de5122c35e with merge 464aedc3ca7d5137fe4f7c3c4b75302931e56aaf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a1c03f31925da058f5bbbb3b2d8e00de5122c35e","merge_sha":"464aedc3ca7d5137fe4f7c3c4b75302931e56aaf"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14000102537/job/39209608866)
Build commit: 464aedc3ca7d5137fe4f7c3c4b75302931e56aaf (`464aedc3ca7d5137fe4f7c3c4b75302931e56aaf`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14000102537/job/39209608866"},"merge_sha":"464aedc3ca7d5137fe4f7c3c4b75302931e56aaf"} -->

rust-timer commented:
Queued 464aedc3ca7d5137fe4f7c3c4b75302931e56aaf with parent a4a11aca5ecf24dfff3c00715641026809951305, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a4a11aca5ecf24dfff3c00715641026809951305&end=464aedc3ca7d5137fe4f7c3c4b75302931e56aaf).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (464aedc3ca7d5137fe4f7c3c4b75302931e56aaf): [comparison URL](https://perf.rust-lang.org/compare.html?start=a4a11aca5ecf24dfff3c00715641026809951305&end=464aedc3ca7d5137fe4f7c3c4b75302931e56aaf&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a4a11aca5ecf24dfff3c00715641026809951305&end=464aedc3ca7d5137fe4f7c3c4b75302931e56aaf&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.2%, 1.1%]   | 8     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.8%, 0.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.5%  | [-0.4%, 1.1%]  | 9     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a4a11aca5ecf24dfff3c00715641026809951305&end=464aedc3ca7d5137fe4f7c3c4b75302931e56aaf&stat=max-rss)
<details>
<summary>Results (primary -3.2%, secondary -3.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.2% | [-3.8%, -2.5%] | 4     |
| Improvements ✅ <br /> (secondary) | -3.2% | [-3.5%, -2.9%] | 2     |
| All ❌✅ (primary)                 | -3.2% | [-3.8%, -2.5%] | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a4a11aca5ecf24dfff3c00715641026809951305&end=464aedc3ca7d5137fe4f7c3c4b75302931e56aaf&stat=cycles:u)
<details>
<summary>Results (primary 1.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [1.1%, 1.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [1.1%, 1.1%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a4a11aca5ecf24dfff3c00715641026809951305&end=464aedc3ca7d5137fe4f7c3c4b75302931e56aaf&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.1%, secondary 0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.1%, 1.3%]   | 9     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.0%, 0.4%]   | 38    |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.1%, -0.0%] | 14    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-0.1%, 1.3%]  | 23    |
</details>

**Bootstrap**: 775.351s -> 776.139s (0.10%)
**Artifact size**: 365.55 MiB -> 365.53 MiB (-0.01%)

thaliaarchi commented:
The max RSS results look promising. Since this aims to reduce memory usage, maybe that overrules the other results? I'm not very experienced at interpreting perf results.

thaliaarchi commented:
I dropped my commit which assumed that `format_args!` pieces cannot overflow `isize::MAX`, since I found a counterexample in https://github.com/rust-lang/rust/issues/138811.

workingjubilee commented:
@workingjubilee you should probably just approve this one yeah? thanks.

workingjubilee commented:
yes I agree @workingjubilee, we should get this show on the road.

@bors r+ rollup

bors commented:
:pushpin: Commit b0a81879c02846f1bded9849c873ee5c9177ecc8 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee b0a81879c02846f1bded9849c873ee5c9177ecc8 -->
<!-- homu: {"type":"Approved","sha":"b0a81879c02846f1bded9849c873ee5c9177ecc8","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
hmm well

@bors rollup=maybe

workingjubilee commented:
@bors rollup=never

There that's what I wanted to put back.

thaliaarchi commented:
Huh. I never actually finished investigating whether the `estimated_capacity` heuristic is still good. I was overcomplicating my measuring strategy and lost interest. But, if the benchmarks look fine to you, it probably is; you can judge them better than me.

bors commented:
:hourglass: Testing commit b0a81879c02846f1bded9849c873ee5c9177ecc8 with merge 01d90ca2b21d46a6cdc700c86b28b8ea6ea8c494...
<!-- homu: {"type":"BuildStarted","head_sha":"b0a81879c02846f1bded9849c873ee5c9177ecc8","merge_sha":"01d90ca2b21d46a6cdc700c86b28b8ea6ea8c494"} -->

Noratrieb commented:
I checked cachegrind on the biggest doc regression of the perf run and found this:
```
>  353,086,439  <alloc::string::String as core::fmt::Write>::write_fmt:???

> -190,119,423  <alloc::raw_vec::RawVecInner<_>>::reserve::do_reserve_and_handle::<alloc::alloc::Global>:???

>  152,807,999  alloc::raw_vec::finish_grow:???

>  148,657,316  alloc::raw_vec::RawVecInner<A>::reserve::do_reserve_and_handle:???

> -101,526,039  alloc::raw_vec::finish_grow::<alloc::alloc::Global>:???

>  -71,185,707  __memcpy_avx_unaligned_erms:???

>  -61,348,606  __rustc::__rust_realloc:???

>   51,174,119  alloc::fmt::format::format_inner:???

>   24,400,284  <alloc::string::String as core::fmt::Write>::write_str:???

>  -15,782,380  alloc::string::String::push:???

>   12,910,951  <alloc::string::String as core::fmt::Write>::write_char:???

>   -4,769,649  __rustc::__rust_alloc:???
```

as usual inlining noise makes it harder to understand, but it does look like it's spending more instructions (which of course don't necessarily have to mean that it's actually slower, but it's a sign) around the code you changed, generally (formatting and appending things to vecs). rustdoc uses a lot of formatting into strings

thaliaarchi commented:
Since this could have a lot of fallout if it's not right, we should pop it from the queue and I should write proper benchmarks, instead of leaning on compile benchmarks. The positive deltas in the Cachegrind results are more than the negative and it seems like it moved where the impact of the reserve happens (hence the different inlining?).

thaliaarchi commented:
I'm going to close it and reopen to cancel the merge.

Zalathar commented:
@bors r-

Zalathar commented:
@bors retry r-

Zalathar commented:
@bors r-

workingjubilee commented:
It seemed Okay. If there was something I missed in my decision-making, it was probably because of the loss in context over time. I definitely don't want to ship something that you aren't happy with yet, obviously.

thaliaarchi commented:
This code is fine, yeah; the possible problem is in code it uses, `estimated_capacity`. I wasn't clear that I didn't consider this ready. Sorry.
